### ISSUE TITLE
[Question]   Dynamic-indexed input causes ~3.9× slowdown vs Triton due to unused threads

### ISSUE THREAD

**Issue by JiatuYan** [tile-ai/tilelang#2458]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### Questions

# `tll.Parallel` scatter-gather: only 16 activated threads cause ~3.9× slowdown vs Triton

## Problem

When a TileLang kernel loads data through a dynamic per-row indirection
(e.g., MoE-style `a[token_ids[i]]` where each M-row reads from a
different source row), the compiler issues `[ThreadSync]` warnings and
the resulting kernel is **~3.9× slower** than an equivalent Triton
kernel on H20 (SM90).

I compiled the kernel into cu file and found that only 16 threads is working.

## Environment

- GPU: H20 (Hopper SM90), CUDA 12.x
- TileLang: 0.1.10, Triton: 3.5.0, PyTorch: 2.9.0

## Compiler warnings

```
[ThreadSync] Hoisting sync from inside if to before if.
Condition is not safe for in-if sync: thread_binding // 16 == 0
```

## Performance

Shape `M=4096 N=512 K=1024 E=8 fp32` (TF32), tiles `BM=16 BN=64 BK=32 NW=4 NS=3`:

```
triton  err=1.33e-01   tilelang err=4.75e-02   (both correct)
     triton    0.1131 ms
   tilelang    0.4395 ms
slowdown = 3.89x
```

When a scatter-store path (`c[token[i], cn] = …`) is also present, the
compiler additionally reports a data-race false positive and inserts
more syncs, making the gap even wider.

## Questions

1. Any suggestions for writing this class of scatter-gather patterns? As I see the deepseek_v4/sparse_attn_fwd_sm90:95-99 is in the same pattern.

## Reproduction

Save as `repro.py` and run `python repro.py`.  Requires `torch`, `triton`,
`tilelang`.

```python
import time
import torch, triton, triton.language as tl
import tilelang.language as tll
from tilelang.jit import jit as tilelang_jit

# ------------------------------------------------------------------ Triton
@triton.jit
def _tri(a_ptr, b_ptr, c_ptr, tok_ptr, exp_ptr, M, N, K,
         BM: tl.constexpr, BN: tl.constexpr, BK: tl.constexpr, GM: tl.constexpr):
    pid = tl.program_id(0)
    np_m, np_n = tl.cdiv(M, BM), tl.cdiv(N, BN)
    grp = GM * np_n
    gid = pid // grp
    gsz = min(np_m - gid * GM, GM)
    pid_m = gid * GM + ((pid % grp) % gsz)
    pid_n = (pid % grp) // gsz
    if pid_m * BM >= M: return

    om = pid_m * BM + tl.arange(0, BM)
    tok = tl.load(tok_ptr + om)
    ob = pid_n * BN + tl.arange(0, BN)
    ok = tl.arange(0, BK)

    a_ptrs = a_ptr + tok[:, None] * K + ok[None, :]
    b_ptrs = b_ptr + tl.load(exp_ptr + pid_m) * N * K + ob[None, :] * K + ok[:, None]

    acc = tl.zeros((BM, BN), tl.float32)
    for _ in range(tl.cdiv(K, BK)):
        acc += tl.dot(tl.load(a_ptrs), tl.load(b_ptrs), allow_tf32=True)
        a_ptrs += BK; b_ptrs += BK

    c_ptrs = c_ptr + om[:, None] * N + ob[None, :]
    tl.store(c_ptrs, acc.to(c_ptr.dtype.element_ty),
             mask=(om[:, None] < M) & (ob[None, :] < N))

# ------------------------------------------------------------------ TileLang
@tilelang_jit()
def _tl(Md, Nd, Kd, Ed, in_d, out_d, sm_d, acc_d, BM, BN, BK, GM, NS, TH):
    npm = tll.dynamic("npm")
    D = {"float32": tll.float32, "float16": tll.float16,
         "bfloat16": tll.bfloat16, "tfloat32": tll.tfloat32}

    @tll.prim_func
    def k(a: tll.Tensor((Md, Kd), D[in_d]),
          b: tll.Tensor((Ed, Nd, Kd), D[in_d]),
          c: tll.Tensor((Md, Nd), D[out_d]),
          tok: tll.Tensor((Md,), "int32"),
          exp: tll.Tensor((npm,), "int32")):
        np_n = tll.ceildiv(Nd, BN)
        npid = tll.ceildiv(Md, BM) * np_n
        nipg = GM * np_n
        nk = tll.ceildiv(Kd, BK)

        with tll.Kernel(npid, threads=TH) as pid:
            gid = pid // nipg
            fm = gid * GM
            gsm = tll.min(npm - fm, GM)
            pm = fm + ((pid % nipg) % gsm)
            pn = (pid % nipg) // gsm
            bsm = pm * BM

            token = tll.alloc_fragment((BM,), "int32")
            for i in tll.Parallel(BM):
                token[i] = tok[bsm + i]
            oe = exp[pm]

            a_s = tll.alloc_shared((BM, BK), D[sm_d])
            b_s = tll.alloc_shared((BN, BK), D[sm_d])
            acc = tll.alloc_fragment([BM, BN], D[acc_d])
            tll.clear(acc)

            for ki in tll.Pipelined(nk, num_stages=NS):
                tll.copy(b[oe, pn * BN: pn * BN + BN, ki * BK: (ki + 1) * BK], b_s)
                for i, j in tll.Parallel(BM, BK):
                    a_s[i, j] = a[token[i], ki * BK + j]
                tll.gemm(a_s, b_s, acc, transpose_B=True)

            ot = tll.alloc_fragment((BM, BN), D[out_d])
            for i, j in tll.Parallel(BM, BN):
                ot[i, j] = tll.Cast(D[out_d], acc[i, j])
            tll.copy(ot, c[bsm: bsm + BM, pn * BN: pn * BN + BN])
    return k

# ------------------------------------------------------------------ Main
def main():
    M, N, K, E = 4096, 512, 1024, 8
    BM, BN, BK, GM, NW, NS = 16, 64, 32, 4, 4, 3
    dt = torch.float32

    a = torch.randn(M, K, dtype=dt, device="cuda")
    b = torch.randn(E, N, K, dtype=dt, device="cuda")
    tok = torch.randint(0, M, (M,), dtype=torch.int32, device="cuda")
    exp = torch.randint(0, E, ((M + BM - 1) // BM,), dtype=torch.int32, device="cuda")
    c0, c1 = torch.zeros(M, N, dtype=dt, device="cuda"), torch.zeros(M, N, dtype=dt, device="cuda")

    def tri():
        grid = (triton.cdiv(M, BM) * triton.cdiv(N, BN),)
        _tri[grid](a, b, c0, tok, exp, M, N, K, BM, BN, BK, GM, num_warps=NW)

    dn = "float32"; sm = "tfloat32"
    comp = _tl(M, N, K, E, dn, dn, sm, "float32", BM, BN, BK, GM, NS, NW * 32)
    tl_launch = lambda: comp(a, b, c1, tok, exp)

    tri(); tl_launch(); torch.cuda.synchronize()

    # correctness
    ac, bc = a.float().cpu(), b.float().cpu()
    cpu = torch.zeros(M, N)
    for m in range(M):
        cpu[m] = ac[int(tok[m].cpu())] @ bc[int(exp[m // BM].cpu())].T
    cpu = cpu.to(dtype=dt)
    print(f"triton  err={(c0.float().cpu()-cpu).abs().max():.2e}   "
          f"tilelang err={(c1.float().cpu()-cpu).abs().max():.2e}")

    # benchmark
    def bench(fn, name, w=10, r=100):
        for _ in range(w): fn()
        torch.cuda.synchronize(); t0 = time.perf_counter()
        for _ in range(r): fn()
        torch.cuda.synchronize()
        ms = (time.perf_counter() - t0) / r * 1e3
        print(f"{name:>9}  {ms:>8.4f} ms")
        return ms
    tri_ms = bench(tri, "triton")
    tl_ms = bench(tl_launch, "tilelang")
    print(f"slowdown = {tl_ms / tri_ms:.2f}x")

if __name__ == "__main__":
    main()
```

The compiled .cu code is as follows:
```
#if defined(_MSC_VER) && !defined(__clang__) && _MSC_VER < 1940
#define _tl_orig_alignas alignas
#define alignas(N) _tl_orig_alignas((N) <= 64 ? (N) : 64)
#include <cuda.h>
#undef alignas
#define alignas _tl_orig_alignas
#endif
#include <tl_templates/cuda/gemm.h>
#include <tl_templates/cuda/copy.h>
#include <tl_templates/cuda/reduce.h>
#include <tl_templates/cuda/ldsm.h>
#include <tl_templates/cuda/threadblock_swizzle.h>
#include <tl_templates/cuda/debug.h>
#ifdef ENABLE_BF16
#include <tl_templates/cuda/cuda_bf16_fallbacks.cuh>
#endif

extern "C" __global__ void kern_kernel(const float* __restrict__ a, int* __restrict__ out, const int* __restrict__ tok);
extern "C" __global__ void __launch_bounds__(128, 1) kern_kernel(const float* __restrict__ a, int* __restrict__ out, const int* __restrict__ tok) {
  int token[1];
  float a_local_cast_1[4];
  int out_local_cast[4];
  token[0] = tok[(((int)threadIdx.x) & 15)];
  if ((((int)threadIdx.x) >> 4) == 0) {
    #pragma unroll
    for (int i = 0; i < 8; ++i) {
      float broadcast_var = 0x0p+0f/*0.000000e+00*/;
      float4 condval;
      if (((0 <= token[0]) && (token[0] < 4096))) {
        condval = *(float4*)(a + ((((int64_t)token[(int64_t)0]) * (int64_t)1024) + (((int64_t)i) * (int64_t)4)));
      } else {
        condval = make_float4(broadcast_var, broadcast_var, broadcast_var, broadcast_var);
      }
      *(float4*)(a_local_cast_1 + 0) = condval;
      int4 __1;
      float4 v_ = *(float4*)(a_local_cast_1 + 0);
      __1.x = (int)(v_.x);
      __1.y = (int)(v_.y);
      __1.z = (int)(v_.z);
      __1.w = (int)(v_.w);
      *(int4*)(out_local_cast + 0) = __1;
      *(int4*)(out + (((((int)threadIdx.x) & 15) * 32) + (i * 4))) = *(int4*)(out_local_cast + 0);
    }
  }
}
```

**Comment by JiatuYan** [comment]
and how can I perform masking (I want to set some lines zero)?

**Comment by Rachmanino** [comment]
I have a improved TileLang kernel that is on par with (or a bit faster than) triton, tested on H200:

```python
"""
Key optimizations:
1. Use float32 shared memory (not tfloat32) to avoid dtype cast overhead
2. Use cp_async for both A scatter-load and B contiguous-load via:
   - tll.Parallel(prefer_async=True, annotations={"parallel_async_without_async_commit_wait": True})
   - tll.async_copy() for B
3. Merge A and B into a single cp_async commit group (ptx_wait_group)
4. Use BN=256 to amortize scatter-load cost over more output columns
5. pass_configs={"tl.disable_safe_memory_legalize": True} to avoid thread guard

"""

import time
import torch, triton, triton.language as tl
import tilelang.language as tll
from tilelang.jit import jit as tilelang_jit

# ------------------------------------------------------------------ Triton
@triton.jit
def _tri(a_ptr, b_ptr, c_ptr, tok_ptr, exp_ptr, M, N, K,
         BM: tl.constexpr, BN: tl.constexpr, BK: tl.constexpr, GM: tl.constexpr):
    pid = tl.program_id(0)
    np_m, np_n = tl.cdiv(M, BM), tl.cdiv(N, BN)
    grp = GM * np_n
    gid = pid // grp
    gsz = min(np_m - gid * GM, GM)
    pid_m = gid * GM + ((pid % grp) % gsz)
    pid_n = (pid % grp) // gsz
    if pid_m * BM >= M: return
    om = pid_m * BM + tl.arange(0, BM)
    tok = tl.load(tok_ptr + om)
    ob = pid_n * BN + tl.arange(0, BN)
    ok = tl.arange(0, BK)
    a_ptrs = a_ptr + tok[:, None] * K + ok[None, :]
    b_ptrs = b_ptr + tl.load(exp_ptr + pid_m) * N * K + ob[None, :] * K + ok[:, None]
    acc = tl.zeros((BM, BN), tl.float32)
    for _ in range(tl.cdiv(K, BK)):
        acc += tl.dot(tl.load(a_ptrs), tl.load(b_ptrs), allow_tf32=True)
        a_ptrs += BK; b_ptrs += BK
    c_ptrs = c_ptr + om[:, None] * N + ob[None, :]
    tl.store(c_ptrs, acc.to(c_ptr.dtype.element_ty),
             mask=(om[:, None] < M) & (ob[None, :] < N))

# ------------------------------------------------------------------ TileLang (optimized)
@tilelang_jit(
    pass_configs={"tl.disable_safe_memory_legalize": True},
)
def _tl(Md, Nd, Kd, Ed, BM, BN, BK, EBM, TH):
    npm = tll.ceildiv(Md, BM)

    @tll.prim_func
    def k(a: tll.Tensor((Md, Kd), tll.float32),
          b: tll.Tensor((Ed, Nd, Kd), tll.float32),
          c: tll.Tensor((Md, Nd), tll.float32),
          tok: tll.Tensor((Md,), "int32"),
          exp: tll.Tensor((tll.ceildiv(Md, EBM),), "int32")):
        np_n = tll.ceildiv(Nd, BN)
        nk = tll.ceildiv(Kd, BK)

        with tll.Kernel(np_n, npm, threads=TH) as (pn, pm):
            bsm = pm * BM

            # Load token indices into shared memory
            token = tll.alloc_shared((BM,), "int32")
            for i in tll.Parallel(BM):
                token[i] = tok[bsm + i]
            oe = exp[bsm // EBM]

            # Use float32 shared memory (not tfloat32!) to enable cp_async
            a_s = tll.alloc_shared((BM, BK), tll.float32)
            b_s = tll.alloc_shared((BN, BK), tll.float32)
            acc = tll.alloc_fragment([BM, BN], tll.float32)
            tll.clear(acc)

            for ki in tll.Pipelined(nk, num_stages=1):
                # Scatter-load A via cp_async (no auto commit/wait)
                for i, j in tll.Parallel(BM, BK, prefer_async=True,
                                         annotations={"parallel_async_without_async_commit_wait": True}):
                    a_s[i, j] = a[token[i], ki * BK + j]

                # Contiguous-load B via async_copy (shares commit group with A)
                tll.async_copy(
                    b[oe, pn * BN: pn * BN + BN, ki * BK: (ki + 1) * BK],
                    b_s,
                )

                # Single wait for both A and B
                tll.ptx_wait_group(0)
                tll.gemm(a_s, b_s, acc, transpose_B=True)

            tll.copy(acc, c[bsm: bsm + BM, pn * BN: pn * BN + BN])
    return k

# ------------------------------------------------------------------ Main
def main():
    M, N, K, E = 4096, 512, 1024, 8
    BM, BN, BK, GM, NW = 16, 64, 32, 4, 4
    dt = torch.float32

    a = torch.randn(M, K, dtype=dt, device="cuda")
    b = torch.randn(E, N, K, dtype=dt, device="cuda")
    tok = torch.randint(0, M, (M,), dtype=torch.int32, device="cuda")
    exp = torch.randint(0, E, ((M + BM - 1) // BM,), dtype=torch.int32, device="cuda")
    c0, c1 = torch.zeros(M, N, dtype=dt, device="cuda"), torch.zeros(M, N, dtype=dt, device="cuda")

    def tri():
        grid = (triton.cdiv(M, BM) * triton.cdiv(N, BN),)
        _tri[grid](a, b, c0, tok, exp, M, N, K, BM, BN, BK, GM, num_warps=NW)

    # TileLang: BN=256 for better amortization of scatter-load cost
    TL_BN = 256
    comp = _tl(M, N, K, E, BM, TL_BN, BK, BM, NW * 32)
    tl_launch = lambda: comp(a, b, c1, tok, exp)

    tri(); tl_launch(); torch.cuda.synchronize()

    # Correctness
    ac, bc = a.float().cpu(), b.float().cpu()
    cpu = torch.zeros(M, N)
    for m in range(M):
        cpu[m] = ac[int(tok[m].cpu())] @ bc[int(exp[m // BM].cpu())].T
    cpu = cpu.to(dtype=dt)
    print(f"triton  err={(c0.float().cpu()-cpu).abs().max():.2e}   "
          f"tilelang err={(c1.float().cpu()-cpu).abs().max():.2e}")
    print(f"triton vs tilelang max diff = {(c0-c1).abs().max().item():.2e}")

    # Benchmark
    def bench(fn, name, w=20, r=500):
        for _ in range(w): fn()
        torch.cuda.synchronize(); t0 = time.perf_counter()
        for _ in range(r): fn()
        torch.cuda.synchronize()
        ms = (time.perf_counter() - t0) / r * 1e3
        print(f"{name:>12}  {ms:>8.4f} ms")
        return ms

    tri_ms = bench(tri, "triton")
    tl_ms = bench(tl_launch, "tilelang")
    print(f"speedup = {tri_ms/tl_ms:.3f}x ({'tilelang FASTER' if tl_ms < tri_ms else 'triton faster'})")

if __name__ == "__main__":
    main()
```
Actually the original triton kernel is buggy to some extent, as it directly consider fp32 inputs as tf32 and pass them to Tensor Core. This explained the **>0.1 abs err**. In contrast, TileLang loads fp32 value and then casts them to tf32 before launching MMAs, which results in significant performance regression. If we impl the kernel in the some way as triton, we can observe the same numeric issues as well as similar performance.

**Comment by JiatuYan** [comment]
Cool! Thanks for your reply. Looks like I missed some key TileLang features and instructions—I gonna try this in my main code.
